### PR TITLE
Check for new added nodes

### PR DESCRIPTION
**Pull request by AbdelrahmanElawady**
### Description

Check for new added nodes periodically and update GPU data when new nodes are added.

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-go/issues/322

### Checklist

- [x] Tests included
- [x] Build pass
- [x] Documentation
- [x] Code format and docstring

### CODE REVIEW

File: grid-proxy/internal/gpuindexer/gpuindexer.go [hunk 1]
```diff
@@ -29,6 +30,7 @@ type NodeGPUIndexer struct {
 	batchSize              int
 	nodesGPUResultsChan    chan []types.NodeGPU
 	nodesGPUBatchesChan    chan []types.NodeGPU
+	newNodeTwinIDChan      chan []uint32
```
Comment (xmonader) [inline grid-proxy/internal/gpuindexer/gpuindexer.go:33]: Why is that a channel of int32 slice? this should keep receiving the latest twinID inserted on the nodes table,  no?

Comment (xmonader) [reply grid-proxy/internal/gpuindexer/gpuindexer.go:33]: ah forget it checks the nodes between the last id you checked and the last id on the table, could be more than one indeed

File: grid-proxy/internal/explorer/db/postgres.go [hunk 2]
```diff
@@ -934,3 +934,9 @@ func (p *PostgresDatabase) UpsertNodesGPU(ctx context.Context, nodesGPU []types.
 	}
 	return nil
 }
+
+func (p *PostgresDatabase) GetLastNodeTwinID(ctx context.Context) (uint32, error) {
```
Comment (xmonader) [inline grid-proxy/internal/explorer/db/postgres.go]: Maybe we should change it to be GetGreaterThanTwinID and use the twinID we checked last? this should get us the list of ids that we can directly send?

Comment (AbdelrahmanElawady) [reply grid-proxy/internal/explorer/db/postgres.go]: Initially, we thought we only need the latest ID and all new nodes will be from last checked to the latest ID however this is not right as twin IDs are shared with other entities so node twin IDs don't have to be incremented one by one. So, this approach is probably better.

File: grid-proxy/internal/gpuindexer/gpuindexer.go [hunk 3]
```diff
@@ -183,10 +195,42 @@ func (n *NodeGPUIndexer) Start(ctx context.Context) {
...
+
 	log.Info().Msg("GPU indexer started")
 
 }
 
+func (n *NodeGPUIndexer) watchNodeTable(ctx context.Context) {
+	ticker := time.NewTicker(newNodesCheckInterval)
+	latestCheckedID, err := n.db.GetLastNodeTwinID(ctx)
+	if err != nil {
+		log.Error().Err(err).Msg("failed to get last node twin id")
```
Comment (xmonader) [inline grid-proxy/internal/gpuindexer/gpuindexer.go:208]: This means the proxy is having a seriou issue where there're no nodes in the table? or could that while the pod is starting for the first time?Is that the reason why we don't panic here?

File: grid-proxy/internal/gpuindexer/gpuindexer.go [hunk 4]
```diff
@@ -183,10 +195,42 @@ func (n *NodeGPUIndexer) Start(ctx context.Context) {
...
+			}
+			nodeTwinIDs := make([]uint32, 0)
+			for _, id := range newIDs {
+				nodeTwinIDs = append(nodeTwinIDs, uint32(id))
+			}
+
+			n.newNodeTwinIDChan <- nodeTwinIDs
+			latestCheckedID = int64(nodeTwinIDs[0])
+		case <-ctx.Done():
+			break
```
Comment (xmonader) [inline grid-proxy/internal/gpuindexer/gpuindexer.go]: should it be return?

File: grid-proxy/internal/gpuindexer/gpuindexer.go (hunk 3 above)
Comment (AbdelrahmanElawady) [reply grid-proxy/internal/gpuindexer/gpuindexer.go:208]: I think there might be non-fatal errors here that would not stop the operation of the indexer, that's why we don't panic. What I can think of is maybe some network issue or db connection error on start. I think we don't need to panic here.

**Review by xmonader — APPROVED** [review]

**Comment by xmonader** [comment]
Thank you 💯

**Pull request merged**